File: Programmers/Lv. 3/단어 변환.py
from collections import deque

def check(str1, str2):
    c = 0
    for i in range(len(str1)):
        if str1[i] != str2[i]:
            c += 1
    if c == 1:
        return True
    else:
        return False
    
def solution(begin, target, words):
    answer = 0
    
    if target not in words:
        return 0
    
    q = deque()
    q.append([begin, 0])
    while q:
        now, cnt = q.popleft()
        if now == target:
            return cnt
        
        for word in words:
            if check(now, word):
                q.append([word, cnt+1])
    
    return answer

File: Programmers/Lv. 3/단어 변환.cpp
#include <string>
#include <vector>
#include <queue>
#include <map>

using namespace std;

bool CheckString(string s1, string s2)
{
    int cnt = 0;
    
    for (int i = 0; i < s1.length(); i++)
    {
        if (s1[i] != s2[i])
        {
            cnt++;
        }
    }
    
    return cnt == 1;
}

int solution(string begin, string target, vector<string> words) {
   
    queue<tuple<string, map<string, bool>, int>> q;
    
    map<string, bool> visited;
    for (string word : words)
    {
        visited[word] = false;
    }
    
    q.push({begin, visited, 0});
    
    while (!q.empty())
    {
        auto tmp = q.front();
        q.pop();
        
        string word = get<0>(tmp);
        auto v = get<1>(tmp);
        int cnt = get<2>(tmp);
        
        if (word == target)
        {
            return cnt;
        }
        
        for (string w : words)
        {
            if (CheckString(word, w))
            {
                if (!v[w])
                {
                    v[w] = true;
                    q.push({w, v, cnt + 1});
                }
            }
        }
    }
    
    return 0;
}